Implement test_provisioning workload

MattLombana opened the pull request:
Implement test_provisioning workload with mgmtsystem
Changes to providers.py:
    - Added functionality for deleting VMs using mgmtsystem library

Changes to mgmt_system/**init**.py:
    - Copied from previous cfme-test repo to extend mgmtsystem to RHEVM infrastructure providers

Changes to test_provsioning.py:
    - Implement test_provisioning workload
    - Time stamp all provisioned vms, and zero pad numbers
    - Collect metrics on total number of provisioned and retired VMs
    - Provision based on template guid rather than name (allows for duplicate named templates)

Changes to workloads.py:
    - Created method get_provisioning_scenarios()

akrzos commented on cfme-performance/conf/cfme_performance.yml:
```diff
@@ -95,6 +95,25 @@ workloads:
         refresh_sleep_time: 600
         total_time: 14400
         grafana_dashboard: cfme-vmware-redhat-system-performance
+  test_provisioning:
```
Lets move this example down to the bottom of the file.  I would like to keep the idea that the hardest to run scenarios are at the bottom of the file thus I believe the current line up should be:
1. idle workloads
2. refresh workloads
3. cap & util workloads
4. SSA workload(s)
5. provisioning workloads

MattLombana replied:
Moved to its own P/r

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,97 @@
+    cleanup_size = scenario['cleanup_size']
+    vm_names = []
+    for i in range(cleanup_size + 1):
+        vm_names.append('TEST_Provision_VM_{}'.format(i))
```
A time stamp pre-fixing the name will help us determine in the event of multiple workloads that involve provisioning which workload "left behind" any vms or is currently provisioning vms so I think our vm name should be something like:

``` python
 vm_name = '{}-provision'.format(time.strftime('%Y%m%d%H%M%S'))
```

I think we should let the value at the end grow to the number of vms we attempted to provision rather than wrap.  Thoughts?

akrzos commented on cfme-performance/utils/providers.py:
```diff
@@ -407,6 +421,95 @@ def refresh_provider_vms_bulk(vm_ids):
+        'action': 'create',
+        'resources': [{
+            'template_fields': {
+                'name': 'perf-provision-rhel7'
```
We will parameterize this and make it part of the provider.

MattLombana replied:
I'm not quite sure what you mean by this. Do you mean that we will add templates to the providers within cfme_performance?

MattLombana commented:
Implemented Changes

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,99 @@
+def test_provisioning(request, scenario):
+    """Runs through provisioning scenarios using the REST API to
+    continously order a catalog item(s) and provision a VM for a specified period of time.
+    Memory Monitor creates graphs and summary at the end of each scenario."""
```
Lets re-word this description of the workload to indicate we are simply using the provisioning_requests api to provision a vm rather than ordering a catalog item.

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,99 @@
+    time_between_provision = scenario['time_between_provision']
+    total_provisioned_vms = 0
+    provisioned_vms = 0
+    provision_order = []
```
This should be defined above the finalizer and passed into the finalizer.  Then the finalizer can pop the remaining vms off the list and attempt to retire them.

akrzos commented on cfme-performance/utils/providers.py:
```diff
@@ -410,6 +487,95 @@ def refresh_provider_vms_bulk(vm_ids):
+
+    if response.status_code != 200:
+        logger.debug(response.text)
+    logger.debug('Provisioned VM {} in: {}s'.format(vm_name,
```
Lets reword that to "queued" since the provisioning takes some time and we wouldn't want to get confused when reading the log output.

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,99 @@
+        monitor_thread.grafana_url = g_url
+        monitor_thread.signal = False
+        monitor_thread.join()
+        retire_provisioned_vms(scenario['vm_names'])
```
We should pass which vms need to be retired rather than passing them in from the scenario.  We can discuss this a bit more tomorrow to make sure we are on the same page WRT that idea.

MattLombana committed:
Implement test_provisioning workload with mgmtsystem

Changes to providers.py:
    - Added functionality for deleting VMs using mgmtsystem library
    - Add provider details to template GUID list to reduce API calls when deleting a VM

    #- Copied from previous cfme-test repo to extend mgmtsystem to RHEVM infrastructure providers

Changes to test_provsioning.py:
    - Implement test_provisioning workload
    - Time stamp all provisioned vms, and zero pad numbers
    - Collect metrics on total number of provisioned and retired VMs
    - Provision based on template guid rather than name (allows for duplicate named templates)

Changes to workloads.py:
    - Created method get_provisioning_scenarios()

Add changes to test_provisioning workload

Changes to cfme_performance.yml
    - Add vlan specification in provider definition

Changes to providers.py:
    - Parameterize vlan in provisioning method
    - Return provider definition as part of template guid

Changes to test_provisioning:
    - Pass vlan into provision_vm()
    - Add scenario data to monitor thread
    - Adjust methods so smem will join threads while cleaning up leftover VMs

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,116 @@
+    scenario_data = {'appliance_ip': cfme_performance['appliance']['ip_address'],
+        'appliance_name': cfme_performance['appliance']['appliance_name'],
+        'test_dir': 'workload-provisioning',
+        'test_name': 'Test Provisioning',
```
I think 'Provisioning' will suffice as a name.  This is used with the top header in the workload report in results directory.

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,116 @@
+        'test_dir': 'workload-provisioning',
+        'test_name': 'Test Provisioning',
+        'appliance_roles': get_server_roles_workload_provisioning(separator=', '),
+        'scenario': {'name': 'default'}}
```
Pass in `scenario` here.  In the idle workloads, there is no scenario so I just passed in a dictionary with the 'name'.

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,116 @@
+        guid_to_provision = next(guid_cycle)
+        provision_order.append([vm_to_provision, guid_to_provision[1]])
+        provision_vm(vm_to_provision, guid_to_provision[0])
+        if provisioned_vms > cleanup_size * providers_num:
```
I think it might make readability a bit better if we just eliminate `providers_num` and use `len(scenario['providers'])`

akrzos commented on cfme-performance/tests/workloads/test_provisioning.py:
```diff
@@ -0,0 +1,116 @@
+        monitor_thread.grafana_url = g_url
+        monitor_thread.signal = False
+        monitor_thread.join()
+        delete_provisioned_vms(vms_to_cleanup)
```
I think if we bump this above `monitor_thread.join()`, it will occur concurrently with the processing of the monitoring thread data.

akrzos commented:
Overall looks great!! Lets make those small changes, squash the commits, merge it and we can test the workload and adjust it as needed afterwards.
